python/lsst/ip/isr/isrMockLSST.py

```diff
@@ -0,0 +1,482 @@
+
+        Returns
+        -------
+        exposure : `lsst.afw.image.Exposure` or `dict`
+            Simulated ISR image data.
+
+        Notes
+        -----
+        This method constructs a "raw" data image.
+        """
+        exposure = self.getExposure()
+
+        # We introduce effects as they happen from a source to the signal,
+        # so the effects go from electrons to ADU.
+        # The ISR steps will then correct these effects in the reverse order.
+        for idx, amp in enumerate(exposure.getDetector()):
+
+            # Get image bbox and data
+            bbox = None
+            if self.config.isTrimmed:
+                bbox = amp.getBBox()
+            else:
+                bbox = amp.getRawDataBBox().shiftedBy(amp.getRawXYOffset())
+
+            ampData = exposure.image[bbox]
+
+            # Sky effects in e-
+            if self.config.doAddSky:
+                self.amplifierAddNoise(ampData, self.config.skyLevel, np.sqrt(self.config.skyLevel))
+
+            if self.config.doAddSource:
+                for sourceAmp, sourceFlux, sourceX, sourceY in zip(self.config.sourceAmp,
+                                                                   self.config.sourceFlux,
+                                                                   self.config.sourceX,
+                                                                   self.config.sourceY):
+                    if idx == sourceAmp:
+                        self.amplifierAddSource(ampData, sourceFlux, sourceX, sourceY)
+
+            # Other effects in e-
+            if self.config.doAddFringe:
+                self.amplifierAddFringe(amp, ampData, np.array(self.config.fringeScale),
+                                        x0=np.array(self.config.fringeX0),
+                                        y0=np.array(self.config.fringeY0))
+
+            if self.config.doAddFlat:
+                if ampData.getArray().sum() == 0.0:
+                    self.amplifierAddNoise(ampData, 1.0, 0.0)
+                u0 = exposure.getDimensions().getX()
+                v0 = exposure.getDimensions().getY()
+                self.amplifierMultiplyFlat(amp, ampData, self.config.flatDrop, u0=u0, v0=v0)
+
+            # ISR effects
+            # 1. Add dark in e- (different from isrMock which does it in ADU)
+            if self.config.doAddDark:
+                self.amplifierAddNoise(ampData,
+                                       self.config.darkRate * self.config.darkTime,
+                                       np.sqrt(self.config.darkRate * self.config.darkTime))
+
+            # 2. Gain normalize  (from e- to ADU)
+            # TODO: DM-??? gain from PTC per amplifier
+            # TODO: DM-??? gain with temperature dependence
+            if self.config.doApplyGain:
+                self.applyGain(ampData, self.config.gain)
+
+            # 3. We either add the bias frame or only the read noise
+            # to the image region in ADU.
```

Comment:
I don't think this comment is correct.  The "bias frame" as we use it as a calibration product is fixed pattern noise, with a mean of ~0.  So I think the adding of noise below is fine, but this is not a bias frame.

Reply:
fixed.

Reply:
My comment is that we need to be extra careful about distinguishing between a bias *frame* which is mean zero fixed pattern noise and the bias *level* which is like 30k and we remove by looking at the overscan levels. (But could be removed with a bias frame I suppose but that’s not how we are going to run lsstcam)

Reply:
Thank you for clarifying - I updated the comment, it also means the biasMockLSST class should actually have a bias level of 0 which I will update.